### DISCUSSION TITLE
Shell plugin example from docs blocks main thread

### DISCUSSION THREAD

**Discussion by StepanZagray** [tauri-apps/tauri#10556] [Q&A] [answered]
Im trying to get "Default Gateway" of local network, basically ip of a router with this function, while using tauri shell plugin:
```rust
pub async fn get_default_gateway(
    app_handle: AppHandle,
) -> Result<String, Box<dyn std::error::Error + Send + Sync>> {
    let shell = app_handle.shell();

    let output = tauri::async_runtime::block_on(async {
        if cfg!(target_os = "windows") {
            // Execute the .bat file using the shell plugin
            shell
                .command("cmd")
                .args(["/C", "src-tauri/get_default_gateway.bat"]) // Ensure the .bat file is in src-tauri
                .output()
                .await
                .unwrap()
        } else {
            // Execute the Unix command using the shell plugin
            shell
                .command("sh")
                .args(["-c", "ip route | grep default | awk '{print $3}'"])
                .output()
                .await
                .unwrap()
        }
    });

    // Check if the command was successful
    if output.status.success() {
        let stdout = String::from_utf8(output.stdout)?.trim().to_string();

        if stdout.is_empty() {
            return Err(Box::new(MeasureLatencyError::DefaultGatewayError));
        }

        Ok(stdout)
    } else {
        // Handle the case where the command fails
        let stderr = String::from_utf8(output.stderr)?;
        eprintln!("Command failed with stderr: {}", stderr);
        Err(Box::new(MeasureLatencyError::DefaultGatewayError))
    }
}
```

I used `tauri::async_runtime::block_on()` as in example in docs, but on execution of this function Im getting panic: 
```
thread 'tokio-runtime-worker' panicked at 
"C:\..\cargo\registry\src\index.crates.io-6f17d22bba15001f\tokio-1.39.2\src\runtime\scheduler\multi_thread\mod.rs:86:9"
Cannot start a runtime from within a runtime. This happens because a function (like `block_on`) attempted to block the current thread while the thread is being used to drive asynchronous tasks.  
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
As I understand `tauri::async_runtime::block_on()` blocks and if I use `tauri::async_runtime::spawn()` in `let output`, `Output` gets wrapped into `Joinhandle` (JoinHandle<Output>) which I don't know have to unwrap(`.unwrap()` doesn't work).

This function is not tauri command, I execute it in `.setup()` in `tauri::async_runtime::spawn()`. I used std::Command before and everything worked fine with simple commands, but for some reason I couldn't get output from commands so I decided to use tauri shell plugin.

**Comment by FabianLars** [comment, marked as answer]
> This function is not tauri command, I execute it in .setup() in tauri::async_runtime::spawn()

Is the setup hook / tauri builder running inside your own tokio runtime (for example if you have an async main fn annotated with `#[tokio::main]`) ?

Anyway, i don't see why you're using block_on in an async function, you can just use `.await` on the futures you want to await/block on

**Reply by StepanZagray** [reply]
It works! Thank you for the answer, don't know why I just didn't `.await`😅
